### PR TITLE
Better handling of points/lines with NaN/Inf and zero length lines for LineLocatePoint and LineInterpolatePoint

### PR DESCRIPTION
**Pull request by mattswoon**
This started out as a quick PR to address the inconsistency I mentioned in #520, but I've also removed some `unwrap` calls I should have been more careful about last time.

Both `LineLocatePoint` and `LineInterpolatePoint` now return NAN or points of NAN where appropriate instead of panicking.

### CODE REVIEW

File: geo/src/algorithm/line_locate_point.rs
Comment (michaelkirk) [inline geo/src/algorithm/line_locate_point.rs:95]: Am I correct that giving the start point of the line gives me the same answer as giving some random point which isn't on the line?

File: geo/src/algorithm/line_interpolate_point.rs
Comment (michaelkirk) [inline geo/src/algorithm/line_interpolate_point.rs:87]: As your test shows, and comment dictates, does this happen if and only if line_string is empty? No other cases?

panic'ing isn't great, but Rust offers some other more expressive tools than sentinel values. I wonder if returning an Option or Result would make more sense.

Would making output an Option or a Result type be better?

File: geo/src/algorithm/line_locate_point.rs
Comment (mattswoon) [reply geo/src/algorithm/line_locate_point.rs:95]: I hope I'm interpreting your question correctly - for a zero length line, yes you'll get zero as the returned fraction for any point in the plane (that's the idea of the test on line 189).

For a non-zero length line, the fold statement below finds the first line segment that is closest to the point, and returns the fractional length of the linestring up to the closest point on that line segment - test on line 182 demonstrates this - the point is equidistant to the first and last line segment, but defers to the first.

Does that make sense/answer your question?

File: geo/src/algorithm/line_interpolate_point.rs
Comment (mattswoon) [reply geo/src/algorithm/line_interpolate_point.rs:87]: I agree - perhaps an enum in the vein of (geo::Closest)[https://docs.rs/geo/0.15.0/geo/enum.Closest.html]?

File: geo/src/algorithm/line_locate_point.rs
Comment (michaelkirk) [reply geo/src/algorithm/line_locate_point.rs:95]: Sorry, my question wasn't very clear.

```
let line_string = line_string!((0.0, 0.0), (10.0, 0.));

// this makes sense to me, we are on the line and 50% along it.
assert_eq!(0.5, line_string.line_locate_point(Point::new(0.0, 5.0)))

// this makes sense to me, we are on the line but 0% along it.
assert_eq!(0.0, line_string.line_locate_point(Point::new(0.0, 0.0)))

// I'm a little surprised that this case is indistinguishable from the previous, since we're not even on the line.
assert_eq!(0.0, line_string.line_locate_point(Point::new(555.0, 555.0))
```

Provided I'm interpreting this correctly, I'm wondering if it would be better to, similar to my other comment, treat the case of "off the line" as some kind of explicit failure, either by returning a Result type or an Option type.

An example of where I could imagine this being surprising:
```
let line_string = line_string!((0.0, 0.0), (10.0, 0.));

assert_eq!(0.0, line_string.line_locate_point(Point::new(0.0, 0.0)))
assert_eq!(0.9, line_string.line_locate_point(Point::new(0.0, 9.0)))
assert_eq!(0.99, line_string.line_locate_point(Point::new(0.0, 9.9)))
assert_eq!(0.999, line_string.line_locate_point(Point::new(0.0, 9.99)))
assert_eq!(1.0, line_string.line_locate_point(Point::new(0.0, 10.0)))

# is this surprising?
assert_eq!(0.0, line_string.line_locate_point(Point::new(0.0, 10.000001)))

# or given this
assert_eq!(0.5, line_string.line_locate_point(Point::new(0.0, 5.0)))
# is this suprising?
assert_eq!(0.0, line_string.line_locate_point(Point::new(0.000001, 5.0)))
```

Granted it's not *that* surprising in a unit test like my example, but if you're building the input point from some sequence of operations where floating point error can accumulate, it seems possible that we'd slip off the line just a bit.

The previous `unwrap` is brutal, but it helped keep the developer from shooting themselves in the foot unknowingly. Maybe explicitly signaling "not on the line" (with an Option or a Result) would be a good compromise.

What do you think @mattswoon?

Comment (mattswoon) [reply geo/src/algorithm/line_locate_point.rs:95]: Right, that makes sense. My intention with these two traits is to help you locate how far one would travel along a line in order to get the point on the line that is _nearest to the given point_, so should return a fraction for any finite point in the plane.  This is the behaviour from postgis, which is why I kept the same names too (https://postgis.net/docs/ST_LineLocatePoint.html).

In your first example, the last assert would actually be `assert_eq!(1.0, line_string.line_locate_point(Point::new(555.0, 555.0))` as the nearest point on `line_string` to `(555, 555)` is the end point.

So my apologies for the confusion, the fraction doesn't "default" to 0.0 for points off the line - below is my crude drawing for what you'd expect for a single line segment.

a single line segment.

For the linestring case, things are slightly more complicated because the point off the line can be equidistant to multiple line segments, and in this case, the fraction return is that for the first, closest line segment. I've got another picture:

linestring

The green and pink dots are equidistant to the first/second and first/third line segments respectively, so the fraction returned is that of the projection to the first line segment. The purple dot is "very close" to the green dot, but closer to the third line segment than the first, so the fraction return is quite a bit larger. But I think this ok.

I'm actually just about to push changes where these will return options, and result in `None` when either the line has nan/infinite coordinates, the point has nan/infinite coordinates for LineLocatePoint, or the fraction is nan for LineInterpolatePoint (interpolating infinitely along a line just gives the end point, and negative-infinitely the start point).

File: geo/src/algorithm/line_interpolate_point.rs
Comment (frewsxcv) [inline geo/src/algorithm/line_interpolate_point.rs:68]: you can use `?` here to flatten this a bit

```suggestion
        match fraction.partial_cmp(&T::zero())? {
            Ordering::Less => return Some(self.start.into()),
            Ordering::Equal => return Some(self.start.into()),
            Ordering::Greater => {
                match fraction.partial_cmp(&T::one())? {
                    Ordering::Greater => return Some(self.end.into()),
                    Ordering::Equal => return Some(self.end.into()),
                    Ordering::Less => {}
                }
            }
        };
```

File: geo/src/algorithm/line_locate_point.rs
Comment (michaelkirk) [reply geo/src/algorithm/line_locate_point.rs]: Thank you for taking the time to explain this.

**Review by michaelkirk — COMMENTED** [review]
hi @mattswoon - thanks for your continued work on this. 

The only *behavioral* concern I had were the changes around `if l.is_finite() {`.  - I wanted you to double check that it makes sense to return None. Maybe it does?

Everything else I commented on were opportunities I saw for improving readability and perf, and I consider them optional if you're getting weary of being in PR. If you'd prefer not to make those changes, I'll follow up after merging.

**Comment by michaelkirk** [comment]
oh and could you please run `cargo fmt` and push up the formatting changes.

**Comment by mattswoon** [comment]
Thanks @michaelkirk this is wonderful. I'll get to these over the next week

File: geo/src/algorithm/line_locate_point.rs
Comment (mattswoon) [reply geo/src/algorithm/line_locate_point.rs:75]: I think the case I had in my head was where you have an infinitely long line - consider
```
let line = line![(x: -inf, y:0), (x:0, y:0)];
let point = (x: -2, y: 2)
```
In this case both `v_dot_sp` and `v_sq` will be `inf`, and I'm not quite sure how to interpret the fraction along an infinite length line anyway.

But I think your case _is_ sensible, so maybe I only need to guard against lines with infinities in them.

File: geo/src/algorithm/line_locate_point.rs
Comment (mattswoon) [reply geo/src/algorithm/line_locate_point.rs:108]: this is what lead me here originally - from #520 

I've tried to defer to postgis, but they seem to choose 1 for zero length Lines and 0 for zero length LineStrings.

I think it's arbitrary, and zero comes before one so ¯\\\_(ツ)\_/¯

File: geo/src/algorithm/line_locate_point.rs
Comment (michaelkirk) [reply geo/src/algorithm/line_locate_point.rs:75]: I doubt my assumption of how things should work are rigorous. I was just being intuitive. How do others (e.g. postgis) handle this situation?

Comment (mattswoon) [reply geo/src/algorithm/line_locate_point.rs:75]: It's not exactly clear from postgis, but it appears they check only whether the line is zero length, and otherwise return the floating point result of `v_dot_sp / v_sq` (maxed/minned to 1/0). This leads to some interesting cases.

```
line: (-1,1) -> (1, 1)
point: (100, inf)
fraction: NaN 
// The dot product of (2, 0) (direction vector) and (101, inf) (point - starting point) is 200 + 0 * inf = 200 + NaN = NaN

line: (-1,1) -> (1, 1 - 0.0000000001)
point: (100, inf)
fraction: 0.0
// The dot product (2, -0.0000000001) and (101, inf) is 200 + (-inf) = -inf

line: (-1,1) -> (1, 1 + 0.0000000001)
point: (100, inf)
fraction: 1.0
// The dot product (2, 0.0000000001) and (101, inf) is 200 + inf = inf
```
So I'm not sure there's much guidance there unfortunately.

Thinking about your example a bit more, I'm having trouble generalizing it - what should the result be for a line `(-10, 0) -> (10, 0)` and a point `(5, inf)`? One might say we should return `0.75` but `(5, inf)` is equally close to `(-10, 0)` and `(10, 0)` if you use a floating-point pythagorus distance - both distances are `inf`. And the dot product of the direction vector and `point - starting point` is NaN.

I'm leaning in the direction that these cases don't really have much meaning, so I think at this stage my preference is either for the following two:

 1. Return `Option<T>` which is `Some` if both point and line have only finite coordinates and `None` otherwise, signifying that locating a point has peculiar semantics (the `Some` variant should only ever contain a finite `T`) (i.e. leaving as is, with some cleaning up).
 2. Return `T` which rolls with the punches of floating points (appears to be the postgis approach)

Either way, a consumer has to think about what to do if their points/lines have non-finite coordinates, but maybe the first option is clearer?

File: geo/src/algorithm/line_interpolate_point.rs
Comment (mattswoon) [reply geo/src/algorithm/line_interpolate_point.rs]: Good question, a bad case of obfuscationism I suspect. I've changed to some more readable if/else statements.

File: geo/src/algorithm/line_interpolate_point.rs
Comment (mattswoon) [reply geo/src/algorithm/line_interpolate_point.rs]: ty for the pro tip, changed these

File: geo/src/algorithm/line_interpolate_point.rs
Comment (mattswoon) [reply geo/src/algorithm/line_interpolate_point.rs]: That is much nicer, updated

File: geo/src/algorithm/line_interpolate_point.rs
Comment (mattswoon) [reply geo/src/algorithm/line_interpolate_point.rs]: tbh I can't tell you why I was doing this allocation in the first place - new approach avoids this

File: geo/src/algorithm/line_interpolate_point.rs
Comment (mattswoon) [reply geo/src/algorithm/line_interpolate_point.rs]: you're exactly right, removed

File: geo/src/algorithm/line_locate_point.rs
Comment (mattswoon) [reply geo/src/algorithm/line_locate_point.rs]: as with the other needless intermediate Vec, I've changed this too. Ty

File: geo/src/algorithm/line_locate_point.rs
Comment (mattswoon) [reply geo/src/algorithm/line_locate_point.rs]: ty, removed Zero, One and PartialOrd

File: geo/src/algorithm/line_locate_point.rs
Comment (mattswoon) [reply geo/src/algorithm/line_locate_point.rs:8]: formatted these a bit nicer

**Comment by michaelkirk** [comment]
Hi @mattswoon - not sure if you're planning to push up any further changes - can you let me know when this is ready for re-review?

**Comment by mattswoon** [comment]
Sorry @michaelkirk yes, this is ready for re-review

File: geo/src/algorithm/line_interpolate_point.rs
Comment (michaelkirk) [inline geo/src/algorithm/line_interpolate_point.rs:93]: hm... what would it mean if we got here? That `total_length.is_nan()`? Can we `debug_assert` that?

Any other cases?

File: geo/src/algorithm/line_interpolate_point.rs
Comment (michaelkirk) [inline geo/src/algorithm/line_interpolate_point.rs:205]: Thanks for testing all these edge cases!

File: geo/src/algorithm/line_locate_point.rs
Comment (michaelkirk) [inline geo/src/algorithm/line_locate_point.rs:50]: Your `$` caused me to learn about KaTex. Nice! I didn't realize rustdoc supported something like that.

File: geo/src/algorithm/line_locate_point.rs
Comment (michaelkirk) [inline geo/src/algorithm/line_locate_point.rs:99]: Both of these methods are much easier for me to follow now, thanks!

File: geo/src/algorithm/line_locate_point.rs
Comment (michaelkirk) [inline geo/src/algorithm/line_locate_point.rs:56]: I think we can remove this `into`

File: geo/src/algorithm/line_locate_point.rs
Comment (michaelkirk) [reply geo/src/algorithm/line_locate_point.rs:70]: After thinking about it for a while... I have no strong opinion. 🤷 

What you have proposed here seems reasonable.

**Review by michaelkirk — COMMENTED** [review]
Some small final nits from me, but otherwise looks good!

File: geo/src/algorithm/line_interpolate_point.rs
Comment (mattswoon) [reply geo/src/algorithm/line_interpolate_point.rs:65]: Added, thanks

File: geo/src/algorithm/line_interpolate_point.rs
Comment (mattswoon) [reply geo/src/algorithm/line_interpolate_point.rs:104]: Also added

File: geo/src/algorithm/line_interpolate_point.rs
Comment (mattswoon) [reply geo/src/algorithm/line_interpolate_point.rs]: so it is, thanks

File: geo/src/algorithm/line_interpolate_point.rs
Comment (mattswoon) [reply geo/src/algorithm/line_interpolate_point.rs:97]: this one made me think - added a debug_assert for whether `fractional_length` is nan, or if the line has no coordinates `self.num_coords() == 0`

**Comment by mattswoon** [comment]
Hi @michaelkirk thanks again for these. I've addressed each in the latest commit.  

One other small change for `line_interpolate_point` - when the fraction compares out of the range [0,1], rather than blindly returning a some of the start or end point, we now call `line_interpolate_point` with fraction = 0 or 1 as appropriate. This helps picks up cases where the line has nans or infs, meaning a fraction in [0,1] would return None, but a fraction outside those points would return a some of the start or end point.

**Comment by michaelkirk** [comment]
bors try 

Hey @mattswoon - this looks great to me! Thanks for iterating on this!

I'm going to leave it open over the weekend in case anyone else wants to take a final look, but otherwise I'll plan to merge monday.

**Review by michaelkirk — APPROVED** [review]

**Comment by michaelkirk** [comment]
Alright, all merged up. Thanks for your work here @mattswoon!

**Pull request merged**
